File: rnamake/lib/RNAMake/src/secondary_structure/ss_tree.cc
//
//  ss_tree.cc
//  RNAMake
//
//  Created by Joseph Yesselman on 5/18/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//

#include <queue>
#include <iostream>

#include "secondary_structure/residue.h"
#include "secondary_structure/ss_tree.h"
#include "secondary_structure/ss_tree_node.h"

namespace sstruct {

struct NodeandIndex {
    SS_NodeDataOP node;
    int index;
    NodeandIndex(SS_NodeDataOP const & n, int i):
        index(i), node(n) {}
};
    
SS_Tree::SS_Tree(
    String const & seq,
    String const & ss):
    tree_(TreeDynamic<SS_NodeDataOP>())
{
    
    ss_ = SecondaryStructure(seq, ss);
    auto pad_residue1 = std::make_shared<Residue>("N", ".", 0, "N", Uuid());
    residues_ = ResidueOPs();
    residues_.push_back(pad_residue1);
    for(auto const & r : ss_.residues()) { residues_.push_back(r); }
    auto pad_residue2 = std::make_shared<Residue>("N", ".", ss_.residues().size(), "N", Uuid());
    residues_.push_back(pad_residue2);
    
    _build_tree();
}

void
SS_Tree::_build_tree() {
    int xb = 1, yb = (int)residues_.size()-2;
    auto ss_chains = std::vector<ChainOP>{ std::make_shared<Chain>(),
                                           std::make_shared<Chain>() };
    auto bounds = ResidueOPs { residues_[xb-1], residues_[yb+1] };
    auto current = std::make_shared<SS_NodeData>( SS_NodeData::SS_Type::SS_SEQ_BREAK,
                                                  ss_chains, bounds);

    std::queue<NodeandIndex> open_nodes;
    open_nodes.push(NodeandIndex(current, -1));
    int index = 0;
    while(! open_nodes.empty()) {
        NodeandIndex current_pair = open_nodes.front();
        open_nodes.pop();
        
        if(current_pair.node->type() == SS_NodeData::SS_Type::SS_HAIRPIN) {
            tree_.add_data(current_pair.node, current_pair.index);
            continue;
        }
        
        xb = _map_back_to_index(current_pair.node->bound_side(0, SS_NodeData::Bound_Side::RIGHT))+1;
        yb = _map_back_to_index(current_pair.node->bound_side(1, SS_NodeData::Bound_Side::LEFT ))-1;
       
        if(xb > yb) {
            index = tree_.add_data(current_pair.node, current_pair.index);
            auto seq_break = _check_from_chain_ends(xb-1, yb+1);
            if(seq_break != nullptr) { tree_.add_data(seq_break, index); }
            continue;
        }
        
        auto next_level = _build_tree_level(xb, yb);
        if (next_level.size() == 0) {
            auto seq_break = _check_from_chain_ends(xb-1, yb+1);
            if(seq_break != nullptr) { tree_.add_data(seq_break); }
        }
        
        if(current_pair.node->type() == SS_NodeData::SS_Type::SS_BULGE) {
            std::vector<SS_NodeDataOP> part_of_nway;
            std::vector<SS_NodeDataOP> not_part_of_nway;
            for(auto const & n : next_level) {
                if(n->type() == SS_NodeData::SS_Type::SS_BULGE ||
                   n->type() == SS_NodeData::SS_Type::SS_HAIRPIN) { part_of_nway.push_back(n); }
                else { not_part_of_nway.push_back(n); }
            }
            next_level = not_part_of_nway;
            //std::cout << xb << " " << yb << " " << part_of_nway.size() << " " << not_part_of_nway.size() << std::endl;
            
            if(part_of_nway.size() > 0) {
                auto ss_chains = current_pair.node->ss_chains();
                auto bounds = current_pair.node->bounds();
                
                for(auto const & n : next_level) {
                    for(int i = 0; i < n->ss_chains().size(); i++) {
                        if(n->ss_chains()[i]->sequence().size() > 0) {
                            ss_chains.push_back(n->ss_chains()[i]);
                        }
                    }
                    
                    if(n->type() == SS_NodeData::SS_Type::SS_BULGE) {
                        int nxb = _map_back_to_index(n->bound_side(0, SS_NodeData::Bound_Side::RIGHT))+1;
                        int nyb = _map_back_to_index(n->bound_side(1, SS_NodeData::Bound_Side::LEFT)) -1;
                        std::vector<SS_NodeDataOP> next_level_2 = _build_tree_level(nxb, nyb);
                        std::cout << next_level_2.size() << std::endl;
                        for(auto const & n2 : next_level_2) { next_level.push_back(n2); }
                    }
                }
                
                auto new_node = std::make_shared<SS_NodeData>(SS_NodeData::SS_Type::SS_NWAY,
                                                              ss_chains, bounds);
                index = current_pair.index;
                current_pair = NodeandIndex(new_node, index);
                
            }
        
        }
        
        index = tree_.add_data(current_pair.node ,current_pair.index);
        for (auto const & n : next_level) {
            open_nodes.push(NodeandIndex(n, index));
        }
        
    }
}
    
    
/*
void
SS_Tree::_build_tree() {
    int xb = 0, yb = (int)seq_.size()-1;
    SS_NodeDataOP current = _assign_new_node(xb, yb);
    int index;
    
    using NodeandIndex = std::pair<SS_NodeDataOP, int>;
    std::queue<NodeandIndex> open_nodes;
    open_nodes.push(NodeandIndex(current, -1));
    
    while(! open_nodes.empty()) {
        NodeandIndex current_pair = open_nodes.front();
        current = current_pair.first;
        
        if(current->type() == SS_NodeData::SS_Type::SS_HAIRPIN) {
            index = tree_.add_data(current, 0, current_pair.second);
            open_nodes.pop();
            continue;
        }
        
        int xb = current->bound_side(0, SS_NodeData::Bound_Side::RIGHT)+1;
        int yb = current->bound_side(1, SS_NodeData::Bound_Side::LEFT) -1;
        
        std::vector<SS_NodeDataOP> next_level = _build_tree_level(xb, yb);
        
        // Check for multi way junctions will be defined as a bulge as a child of another bulge
        if(current->type() == SS_NodeData::SS_Type::SS_BULGE) {
            std::vector<SS_NodeDataOP> part_of_nway;
            std::vector<SS_NodeDataOP> not_part_of_nway;
            for(auto const & n : next_level) {
                if(n->type() == SS_NodeData::SS_Type::SS_BULGE ||
                   n->type() == SS_NodeData::SS_Type::SS_HAIRPIN) { part_of_nway.push_back(n); }
                else { not_part_of_nway.push_back(n); }
            }
            next_level = not_part_of_nway;
            
            if(part_of_nway.size() > 0) {
                Strings seq;
                std::vector<Ints> bounds;
                
                seq.push_back(current->seqs()[0]);
                bounds.push_back(current->bounds(0));

                seq.push_back(current->seqs()[1]);
                bounds.push_back(current->bounds(1));
                
                for(auto const & n : part_of_nway) {
                    Strings child_seq = n->seqs();
                    for(int i = 0; i < child_seq.size(); i++) {
                        if(child_seq[i].size() > 0) {
                            seq.push_back(child_seq[i]);
                            bounds.push_back(n->bounds(i));
                        }
                    }
                    
                    if(n->type() == SS_NodeData::SS_Type::SS_BULGE) {
                        int nxb = n->bound_side(0, SS_NodeData::Bound_Side::RIGHT)+1;
                        int nyb = n->bound_side(1, SS_NodeData::Bound_Side::LEFT) -1;
                        std::vector<SS_NodeDataOP> next_level_2 = _build_tree_level(nxb, nyb);
                        for(auto const & n2 : next_level_2) { next_level.push_back(n2); }
                    }
                }
                
                
                current = std::make_shared<SS_NodeData>(seq, SS_NodeData::SS_Type::SS_NWAY, bounds);
            }
            
        }
        
        index = tree_.add_data(current, 0, current_pair.second);
        open_nodes.pop();
        
        for(auto const & n : next_level) {
            open_nodes.push(NodeandIndex(n, index));
        }
    }
    
}

*/

std::vector<SS_NodeDataOP>
SS_Tree::_build_tree_level(
    int xb,
    int yb) {
    
    std::vector<SS_NodeDataOP> next_level;
    while(xb <= yb) {
        if(xb == yb && residues_[xb]->dot_bracket() != ".") { break; }
        auto child = _assign_new_node(xb, yb);
        xb = _map_back_to_index(child->bound_side(1, SS_NodeData::Bound_Side::RIGHT))+1;
        next_level.push_back(child);
    }
    
    return next_level;
    
}
    
SS_NodeDataOP
SS_Tree::_assign_new_node(
    int xb,
    int yb) {
    
    std::vector<ResidueOPs> ss_chain_res(2);
    ResidueOPs bounds(2);
    bounds[0] = residues_.at(xb-1);
    bounds[1] = residues_.at(yb+1);
    int hairpin = 0;
    
    if(residues_.at(xb)->dot_bracket() == ".") {
        int end_x = _get_dot_bounds(xb, 0);
        for(int i = xb; i <= end_x; i++) {
            ss_chain_res[0].push_back(residues_[i]);
        }
        if(end_x == yb) { hairpin = 1; }
    }

    if(residues_.at(yb)->dot_bracket()== "." && hairpin == 0) {
        int end_y = _get_dot_bounds(yb, 1);
        for(int i = end_y; i <= yb; i++) {
            ss_chain_res[1].push_back(residues_[i]);
        }
    }
    
    SS_NodeData::SS_Type type;
    if(ss_chain_res[0].size() == 0 && ss_chain_res[1].size() == 0) {
        int pair = _get_brack_pair(xb);
        ss_chain_res[0].push_back(residues_.at(xb));
        ss_chain_res[1].push_back(residues_.at(pair));
        type = SS_NodeData::SS_Type::SS_BP;
        if(residues_.at(xb)->dot_bracket() == "[") {
            type = SS_NodeData::SS_Type::SS_PSEUDO_BP;
        }
    }
    
    else if(hairpin) {
        type = SS_NodeData::SS_Type::SS_HAIRPIN;
    }

    else {
        type = SS_NodeData::SS_Type::SS_BULGE;
    }

    ChainOPs ss_chains(2);
    ss_chains[0] = std::make_shared<Chain>(ss_chain_res[0]);
    ss_chains[1] = std::make_shared<Chain>(ss_chain_res[1]);
    
    auto current = std::make_shared<SS_NodeData>(type, ss_chains, bounds);
    
    return current;

}

    
int
SS_Tree::_get_brack_pair(
    int pos) {
    
    int bracket_count = 0;
    int i = -1;
    for(auto const & r : residues_) {
        i++;
        if( i < pos) { continue; }
        if (r->dot_bracket() == "(" || r->dot_bracket() == "[") { bracket_count++; }
        if (r->dot_bracket() == ")" || r->dot_bracket() == "]") { bracket_count--; }
        if(bracket_count == 0) { return i; }
    }
    
    throw SecondaryStructureException("could not find bracket pair");

}

int
SS_Tree::_get_dot_bounds(
    int pos,
    int reverse) {
    
    
    int incr = 1;
    if(reverse) { incr = -1; }
    
    while(1) {
        if(pos == 0 || pos == residues_.size()) { break; }
        if(residues_[pos]->dot_bracket() == ".") { pos += incr; }
        else{
            pos -= incr;
            break;
        }
    }
    return pos;
}
    
int
SS_Tree::_map_back_to_index(
    ResidueOP const & r) {
    
    int i = 0;
    try {
        i = (int)(std::find(residues_.begin(), residues_.end(), r) - residues_.begin());
    }
    catch(...) {
        throw SecondaryStructureException("unexpected error, cannot find residue in internal array, in SS_Tree, this should not happen!");
    }
    
    return i;
    
}
    
SS_NodeDataOP
SS_Tree::_check_from_chain_ends(
    int xb,
    int yb) {
    
    if(! _is_res_end_of_chain(residues_[xb]->num()) &&
       ! _is_res_end_of_chain(residues_[yb]->num())) {
        return nullptr;
    }
    
    auto fake_res_1 = std::make_shared<Residue>("&", "&", residues_[xb]->num(), "N", Uuid());
    auto fake_res_2 = std::make_shared<Residue>("&", "&", residues_[yb]->num(), "N", Uuid());
    auto ss_chain_res = std::vector<ResidueOPs>{ {fake_res_1}, {fake_res_2} };
    auto ss_chains = ChainOPs { std::make_shared<Chain>(ss_chain_res[0]),
                                std::make_shared<Chain>(ss_chain_res[1]) };
    auto bounds = ResidueOPs { residues_[xb-1], residues_[yb+1] };
    auto current = std::make_shared<SS_NodeData>(SS_NodeData::SS_Type::SS_SEQ_BREAK,
                                                 ss_chains, bounds);
    return current;
}

int
SS_Tree::_is_res_end_of_chain(
    int num) {
    
    for (auto const & c : ss_.chains()) {
        if(c->last()->num() == num) {
            return 0;
        }
    }
    return 1;
}

    
} //sstruct

File: rnamake/ss_tree.py
import tree
import Queue
import math
import secondary_structure
import uuid
from collections import namedtuple

def compare_ss_tree(ss_tree1, ss_tree2):

    bp_score = {
        "G+C-A+U" : 0.25,
        "A+U-G+C" : 0.25,
        "G+C-U+A" : 0.50,
        "U+A-G+C" : 0.50,
        "C+G-A+U" : 0.50,
        "A+U-C+G" : 0.50,
        "C+G-U+A" : 0.25,
        "U+A-C+G" : 0.25
    }

    n_score = {
        "G-A" : 1,
        "G-U" : 2,
        "G-C" : 3,
        "A-G" : 1,
        "A-C" : 2,
        "A-U" : 3,
        "C-U" : 1,
        "C-A" : 2,
        "C-G" : 3,
        "U-C" : 1,
        "U-A" : 3,
        "U-G" : 2
    }

    score = 0

    for n1 in ss_tree1:
        n2 = ss_tree2.get_node(n1.index)

        if n1.data.type == SS_Type.SS_BP:
            if n1.data.seq() ==  n2.data.seq():
                continue
            else:
                str_id = n1.data.seq() + "-" + n2.data.seq()
                if str_id not in bp_score:
                    score += 0.75
                else:
                    score += bp_score[str_id]

        else:
            spl1 = n1.data.seq().split("+")
            spl2 = n2.data.seq().split("+")

            if len(spl1) != len(spl2):
                return 1000

            for i in range(len(spl1)):
                if spl1[i] == spl2[i]:
                    continue

                min = len(spl1[i])
                if len(spl2[i]) < min:
                    min = len(spl2[i])
                score += (math.fabs(len(spl1[i]) - len(spl2[i])))*2

                for j in range(min):
                    if spl1[i][j] == spl2[i][j]:
                        continue
                    c_n_score = n_score[ spl1[i][j] + "-" + spl2[i][j]]
                    score += c_n_score

    return score


class SS_Tree(object):
    def __init__(self, sequence=None, dot_bracket=None, ss=None):
        self.tree = tree.TreeDynamic()

        if sequence is None and dot_bracket is None and ss is None:
            raise ValueError("must supply either sequence and dot_bracket or " +\
                             "secondary structure object")

        if   sequence is not None and dot_bracket is not None:
            self.ss = secondary_structure.SecondaryStructure(sequence, dot_bracket)
        elif ss is not None:
            self.ss = ss
        else:
            raise ValueError("must supply either sequence and dot_bracket or " +\
                             "secondary structure object")

        self.residues = self.ss.residues()
        pad_residue1 = secondary_structure.Residue("N", ".", 0, "N", uuid.uuid1())
        self.residues.insert(0, pad_residue1)
        pad_residue2 = secondary_structure.Residue("N", ".", len(self.residues),
                                                   "N", uuid.uuid1())
        self.residues.append(pad_residue2)
        self._build_tree()

    def get_node(self, i):
        return self.tree.get_node(i)

    def __len__(self):
        return len(self.tree)

    def __iter__(self):
        self.tree.__iter__()
        return self

    def next(self):
        return self.tree.next()

    def _check_from_chain_ends(self, xb, yb):
        if self._is_res_end_of_chain(self.residues[xb].num) or \
           self._is_res_end_of_chain(self.residues[yb].num):
            ss_chains = [secondary_structure.Chain(),
                         secondary_structure.Chain()]
            fake_res1 = secondary_structure.Residue('&', '&', self.residues[xb].num,
                                                    "N", uuid.uuid1())
            fake_res2 = secondary_structure.Residue('&', '&', self.residues[yb].num,
                                                    "N", uuid.uuid1())
            ss_chains[0].residues.append(fake_res1)
            ss_chains[1].residues.append(fake_res2)
            current = SS_NodeData(SS_Type.SS_SEQ_BREAK, ss_chains,
                                  [ self.residues[xb-1], self.residues[yb+1]])
            return current
        return None

    def _build_tree(self):
        xb, yb = 1, len(self.residues)-2
        ss_chains = [secondary_structure.Chain(),
                    secondary_structure.Chain()]
        current = SS_NodeData(SS_Type.SS_SEQ_BREAK, ss_chains,
                              [ self.residues[xb-1], self.residues[yb+1]])
        NodeandIndex = namedtuple('NodeandIndex', 'node index')
        open_nodes = Queue.Queue()
        open_nodes.put(NodeandIndex(current, -1))
        while not open_nodes.empty():
            current_pair = open_nodes.get()

            if current_pair.node.type == SS_Type.SS_HAIRPIN:
                index = self.tree.add_data(current_pair.node, current_pair.index)
                continue

            xb =  self._map_back_to_index(current_pair.node.bound_side(0, Bound_Side.RIGHT))+1
            yb =  self._map_back_to_index(current_pair.node.bound_side(1, Bound_Side.LEFT)) -1


            if xb > yb:
                index = self.tree.add_data(current_pair.node, current_pair.index)
                seq_break = self._check_from_chain_ends(xb-1, yb+1)
                if seq_break is not None:
                    self.tree.add_data(seq_break, index)

                continue

            next_level = self._build_tree_level(xb, yb)

            if len(next_level) == 0:
                seq_break = self._check_from_chain_ends(xb-1, yb+1)
                if seq_break is not None:
                    next_level.append(seq_break)

            if current_pair.node.type == SS_Type.SS_BULGE:
                part_of_nway, not_part_of_nway = [], []
                for n in next_level:
                    if n.type == SS_Type.SS_BULGE or n.type == SS_Type.SS_HAIRPIN:
                        part_of_nway.append(n)
                    else:
                        not_part_of_nway.append(n)
                next_level = not_part_of_nway

                if len(part_of_nway) > 0:
                    ss_chains = current_pair.node.ss_chains
                    bounds = current_pair.node.bounds

                    for n in part_of_nway:
                        for i in range(len(n.ss_chains)):
                            if len(n.ss_chains[i].sequence()) > 0:
                                ss_chains.append(n.ss_chains[i])

                        if n.type == SS_Type.SS_BULGE:
                            nxb = self._map_back_to_index(n.bound_side(0, Bound_Side.RIGHT))+1
                            nyb = self._map_back_to_index(n.bound_side(1, Bound_Side.LEFT)) -1
                            next_level_2 = self._build_tree_level(nxb, nyb)
                            for n2 in next_level_2:
                                next_level.append(n2)

                    new_node = SS_NodeData(SS_Type.SS_NWAY, ss_chains, bounds)
                    index = current_pair.index
                    current_pair = NodeandIndex(new_node, index)

            index = self.tree.add_data(current_pair.node, current_pair.index)
            for n in next_level:
                open_nodes.put(NodeandIndex(n, index))

    def _build_tree_sub(self, xb, yb):
        pass

    def _build_tree_level(self, xb, yb):
        next_level = []

        while xb <= yb:
            if xb == yb and self.residues[xb].dot_bracket != ".":
                break
            child = self._assign_new_node(xb, yb)
            xb =  self._map_back_to_index(child.bound_side(1, Bound_Side.RIGHT))+1
            next_level.append(child)
        return next_level

    def _assign_new_node(self, xb, yb):
        ss_chains = [secondary_structure.Chain(),
                     secondary_structure.Chain()]
        bounds = [self.residues[xb-1], self.residues[yb+1]]
        hairpin = 0
        current = None

        if self.residues[xb].dot_bracket == '.':
            end_x = self._get_dot_bounds(xb, 0)
            for i in range(xb, end_x+1):
                ss_chains[0].residues.append(self.residues[i])
            if end_x == yb:
                hairpin = 1

        if self.residues[yb].dot_bracket == '.' and hairpin == 0:
            end_y = self._get_dot_bounds(yb, 1)
            for i in range(end_y, yb+1):
                ss_chains[1].residues.append(self.residues[i])

        type = None
        if len(ss_chains[0].residues) == 0 and len(ss_chains[1].residues) == 0:
            pair = self._get_bracket_pair(xb)
            ss_chains[0].residues.append(self.residues[xb])
            ss_chains[1].residues.append(self.residues[pair])
            type = SS_Type.SS_BP
            if(self.residues[xb].dot_bracket == '['):
               type = SS_Type.SS_PSEUDO_BP

        elif hairpin:
            type = SS_Type.SS_HAIRPIN

        else:
            type = SS_Type.SS_BULGE

        current = SS_NodeData(type, ss_chains, bounds)

        return current

    def _get_bracket_pair(self, num):
        """
        finds the position of ")" pair to the current "(" in the secondary
        structure array

        :params pos: position in secondary structure of current "("
        :type pos: int
        :params ss: secondary structure
        :type ss: List
        :returns: int -- the position of the paired ")" in list ss
        :raises: ValueError

        >>>print get_bracket_pair(0, [(,(, ., ), )])
        4
        """
        bracket_count = 0
        for i, r in enumerate(self.residues):
            if i < num:
                continue
            if r.dot_bracket == "(":
                bracket_count += 1
            if r.dot_bracket == ")":
                bracket_count -= 1
                if bracket_count == 0:
                    return i

        raise ValueError("cannot find pair")

    def _get_dot_bounds(self, pos,reverse=0):
        """
        finds the bounds of a current dot stretch

        :params pos: position in secondary structure of current "."
        :type pos: int
        :params ss: secondary structure
        :type ss: List
        :params reverse: direction of search::
            0 -- postive, going left to right in the array
            1 -- negative, going right to left in the array
        :type reverse: int
        :state reverse: 0
        :returns: int -- the position of last "." in stretch

        >>>print get_dot_bounds(1,[(, ., ., ., )])
        3

        >>>print get_dot_bounds(3,[(, ., ., ., )],reverse=1)
        1
        """

        incr = 1
        if reverse == 1:
            incr = -1

        while 1:
            if pos == 0 or pos == len(self.residues):
                break

            if self.residues[pos].dot_bracket == ".":
                pos += incr
            else:
                pos -= incr
                break



        return pos

    def _is_res_end_of_chain(self, num):
        for c in self.ss.chains:
            if c.last().num == num:
                return 1
        return 0

    def _map_back_to_index(self, r):
        try:
            i = self.residues.index(r)
            return i
        except:
            raise ValueError("cannot find residue " + r )

class SS_Type(object):
   SS_BP        = 0
   SS_BULGE     = 1
   SS_HAIRPIN   = 2
   SS_NWAY      = 3
   SS_PSEUDO_BP = 4
   SS_SEQ_BREAK = 5
   SS_MULTI     = 6

class Bound_Side(object):
    LEFT  = 0
    RIGHT = 1

class SS_NodeData(object):
    def __init__(self, type, ss_chains, bounds):
        self.ss_chains, self.type, self.bounds = ss_chains, type, bounds

    def bound_side(self, pos, side):
        ss_chain = self.ss_chains[pos]
        if len(ss_chain.residues) == 0:
            return self.bounds[pos]

        if side == Bound_Side.LEFT:
            return ss_chain.residues[0]
        else:
            return ss_chain.residues[-1]

    def what(self):
        if   self.type == SS_Type.SS_BP:        return "SS_BP"
        elif self.type == SS_Type.SS_BULGE:     return "SS_BULGE"
        elif self.type == SS_Type.SS_HAIRPIN:   return "SS_HAIRPIN"
        elif self.type == SS_Type.SS_NWAY:      return "SS_NWAY"
        elif self.type == SS_Type.SS_PSEUDO_BP: return "SS_PSEUDO_BP"
        elif self.type == SS_Type.SS_SEQ_BREAK: return "SS_SEQ_BREAK"
        else: raise ValueError("unknown SS_TYPE")

    def sequence(self):
        seqs = [c.sequence() for c in self.ss_chains]
        return "&".join(seqs)

